src/editor/css_sort.rs

```diff
@@ -0,0 +1,193 @@
+
+        assert_eq!(html, expected);
+    }
+
+    #[test]
+    fn simple_example() {
+        let input = "<p>A</p><p>C</p><p>B</p>";
+        let expected = "<p>A</p><p>B</p><p>C</p>";
+        let sort_by = CssSort {
+            selector: Selector::parse("p").unwrap(),
+            sort_by: Vec::new(),
+            reverse: false,
+        };
+        case(&sort_by, input, expected);
+    }
+
+    #[test]
+    fn reverse() {
+        let input = "<p>A</p><p>C</p><p>B</p>";
+        let expected = "<p>C</p><p>B</p><p>A</p>";
+        let sort_by = CssSort {
+            selector: Selector::parse("p").unwrap(),
+            sort_by: Vec::new(),
+            reverse: true,
+        };
+        case(&sort_by, input, expected);
+    }
+
+    #[test]
+    fn sort_by() {
+        let input = r#"<article><h3>A</h3><a id="Y">Bla</a></article><article><h3>B</h3><a id="X">Bla</a></article>"#;
+        let expected = r#"<article><h3>B</h3><a id="X">Bla</a></article><article><h3>A</h3><a id="Y">Bla</a></article>"#;
+        let sort_by = CssSort {
+            selector: Selector::parse("article").unwrap(),
+            sort_by: vec![Editor::CssSelect(Selector::parse("a").unwrap())],
+            reverse: false,
+        };
+        case(&sort_by, input, expected);
+    }
+
+    #[test]
+    fn sort_by_same_key_keeps_order() {
+        let input = r#"<article><h3>C</h3><a id="X">Bla</a></article><article><h3>A</h3><a id="X">Bla</a></article>"#;
+        let expected = r#"<article><h3>C</h3><a id="X">Bla</a></article><article><h3>A</h3><a id="X">Bla</a></article>"#;
+        let sort_by = CssSort {
+            selector: Selector::parse("article").unwrap(),
+            sort_by: vec![Editor::CssSelect(Selector::parse("a").unwrap())],
+            reverse: false,
+        };
+        case(&sort_by, input, expected);
+    }
+
+    #[test]
+    fn sorting_toplevel_keeps_children_unsorted() {
+        let input = "<div><p>D</p><p>A</p></div><div><p>C</p><p>B</p></div>";
+        let expected = "<div><p>C</p><p>B</p></div><div><p>D</p><p>A</p></div>";
+        let sort_by = CssSort {
+            selector: Selector::parse("div").unwrap(),
+            sort_by: Vec::new(),
+            reverse: false,
+        };
+        case(&sort_by, input, expected);
+    }
+
+    #[test]
+    fn sorting_bottomlevel_keeps_parents_unsorted() {
+        let input = "<div><p>D</p><p>A</p></div><div><p>C</p><p>B</p></div>";
+        let expected = "<div><p>A</p><p>D</p></div><div><p>B</p><p>C</p></div>";
+        let sort_by = CssSort {
+            selector: Selector::parse("p").unwrap(),
+            sort_by: Vec::new(),
+            reverse: false,
+        };
+        case(&sort_by, input, expected);
+    }
+
+    /// Currently, this is done in a simplified way and maybe should be improved in the future.
+    /// For now this documents that its not working perfectly.
+    #[test]
+    fn sort_with_other_elements() {
```

Comment:
I already forgot what the issue was and this comment isn't helping that much.
maybe add something like "Not needed html tags are collectively appended in an undefined order"?
This should / could also be more specific in the README which just says: "Other elements not matching are kept."

Reply:
This test case shows pretty well what actually happens.

> undefined order

Its deterministic but unintuitive. For every use case I had so far it didn't really matter, so I'm not sure its worth looking that much into it. Maybe in the future someone opens an issue with an explanation where its more relevant. Until then, I wouldn't keep much attention into this simplified but not perfect behavior.

Reply:
I see. I agree the test case shoes what happens. But it does not tell the details, that you just commented.
An undefined order can be deterministic in a sense that it is implementation defined. I agree that this is not something to put effort in how ever, the comment is still bogus imo.
Especially the first line: `Currently, this is done in a simplified way and maybe should be improved in the future.`
Currently? Should be improved in the future? Do you plan it? etc. etc.
My proposal:
`Elements that are not of interest during sorting are appended to the output.`
`They are appended in a deterministic but undefined order, to ease use with git / editors.`